src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -12,325 +13,110 @@ namespace System
+                // fit an entire vector's worth of data. Instead of falling back to a loop, we'll write
+                // a vector at the very end of the buffer. This may involve overwriting previously
+                // populated data, which is fine since we're splatting the same value for all entries.
+                // There's no need to perform a length check here because we already performed this
+                // check before entering the vectorized code path.
 
-            // This is separated out into a different variable, so the i + 16 addition can be
-            // performed at the start of the pipeline and the loop condition does not have
-            // a dependency on the writes.
-            nuint counter;
+                Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref b, byteLength - (nuint)Vector<byte>.Count), vector);
 
-            do
-            {
-                counter = i + 16;
+                // And we're done!
 
-                // This loop looks very costly since there appear to be a bunch of temporary values
-                // being created with the adds, but the jit (for x86 anyways) will convert each of
-                // these to use memory addressing operands.
+                return;
+            }
 
-                // So the only cost is a bit of code size, which is made up for by the fact that
-                // we save on writes to b.
+        CannotVectorize:
 
-#if TARGET_64BIT
-                Unsafe.As<byte, long>(ref Unsafe.AddByteOffset<byte>(ref b, i)) = 0;
-                Unsafe.As<byte, long>(ref Unsafe.AddByteOffset<byte>(ref b, i + 8)) = 0;
-#else
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i)) = 0;
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i + 4)) = 0;
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i + 8)) = 0;
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i + 12)) = 0;
-#endif
+            // If we reached this point, we cannot vectorize this data, or there are too few
+            // elements for us to vectorize. Fall back to an unrolled loop.
 
-                i = counter;
+            nuint i = 0;
 
-                // See notes above for why this wasn't used instead
-                // i += 16;
-            }
-            while (counter <= end);
+            // Write 8 elements at a time
 
-            if ((byteLength & 8) != 0)
+            if (byteLength >= 8)
             {
-#if TARGET_64BIT
-                Unsafe.As<byte, long>(ref Unsafe.AddByteOffset<byte>(ref b, i)) = 0;
-#else
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i)) = 0;
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i + 4)) = 0;
-#endif
-                i += 8;
+                nuint stopLoopAtOffset = byteLength & ~(nuint)7;
+                do
+                {
+                    Unsafe.Add(ref b, (nint)i + 0) = 0;
```

Comment:
This should do better than writing one byte at a time....

Reply:
I'll change, but this should not be an issue for jit after https://github.com/dotnet/runtime/pull/92852 and https://github.com/dotnet/runtime/pull/92939 🙂 

E.g.:

![image](https://github.com/dotnet/runtime/assets/523221/bbe54fc9-38c0-4624-8fd0-cb87efff74b1)

The whole codegen on Windows-x64-AVX is  https://gist.github.com/EgorBo/a62071d880c04f59998464013f78e1f3

Reply:
A small issue in the codegen is that zero vector is propagated everywhere by value so we end up doing redundant `vxorps   ymm0, ymm0, ymm0`, but it's a known issue we currently don't have a good solution for

Reply:
Cool! 

We may want to handle the last misaligned element via overlapping write, similar to how the vectorized path does it.

Reply:
Just realized that I posted codegen with Avx512F=0 (disabled), when AVX512 is presented it's also merges two V256 stores in the main loop: https://gist.github.com/EgorBo/f404b7907781059c7b7beecd4f9c42d9